elfpy/markets/hyperdrive/hyperdrive_market.py

```diff
@@ -642,3 +640,169 @@ def apply_checkpoint(self, checkpoint_time: float, share_price: float) -> float:
+        recovered_margin = (
+            shares * self.market_state.withdraw_capital / self.market_state.withdraw_shares_ready_to_withdraw
+        )
+
+        # The user gets interest equal to their percent of the withdraw pool times the withdraw pool
+        # interest
+        recovered_interest = (
+            shares * self.market_state.withdraw_interest / self.market_state.withdraw_shares_ready_to_withdraw
+        )
+
+        # Update the pool state
+        # Note - Will revert here if not enough margin has been reclaimed by checkpoints or by
+        #  position closes
+        market_deltas.withdraw_shares_ready_to_withdraw -= shares
+        market_deltas.withdraw_capital -= recovered_margin
+        market_deltas.withdraw_interest -= recovered_interest
+
+        # Withdraw for the user
+        base_proceeds = self._withdraw(recovered_margin + recovered_interest, as_underlying)
+        # TODO: figure out how to keep track of hyperdrive's base asset amount.  market_deltas has
+        # a d_base_asset, but that is used to update the share_reserves :/.
+        # market_deltas.d_base_asset -= base_proceeds
+        wallet_deltas.balance.amount += base_proceeds
+
+        # Enforce min user outputs
+        if min_output > base_proceeds:
+            raise errors.OutputLimit
+
+        return market_deltas, wallet_deltas
+
+    def _withdraw(self, shares: float, as_underlying: bool) -> float:
+        r"""
+        Calculates the amount of base to withdraw for a given amount of shares.
+
+        Parameters
+        ----------
+        shares: float
+            The withdrawal shares to redeem.
+        as_underlying: bool
+            If true, the user is paid in underlying, if false the contract transfers in yield source
+            directly. Note - for some paths one choice may be disabled or blocked.
+
+        Returns
+        -------
+        float
+          The withdraw_value and share_price as a tuple.
+        """
+
+        # This yield source doesn't accept the underlying since it's just base.
+        if not as_underlying:
+            raise errors.UnsupportedOption
+
+        # TODO: add step to accrue interest
+
+        # Get the amount of base to transfer.
+        amount_withdrawn = shares * self.market_state.share_price
+        return amount_withdrawn
+
+    def _free_margin(
+        self, freed_capital: float, max_capital: float, interest: float
+    ) -> hyperdrive_actions.MarketDeltas:
+        """
+        Moves capital into the withdraw pool and marks shares ready for withdraw.
+
+        Parameters
+        ----------
+        freed_capital: float
+            The amount of capital to add to the withdraw pool, must not be more than the max capital.
+        max_capital: float
+            The margin which the LP used to back the position which is being closed.
+        interest: float
+            The interest earned by this margin position, fixed interest for LP shorts and variable for longs.
```

Comment:
is this supposed to be "(add & remove) LP and shorts" or is "LP shorts" a thing? Is it the opposing trade to an agent opening a long?

Reply:
LP was a typo there, copy pasta'd from the solidity.  it now reads 'fixed interest for shorts and variable for longs.'